foreman/client.py

```diff
@@ -356,14 +372,43 @@ def __init__(self, url, auth=None, version=None, api_version=None,
         # Instantiate plugins
         self.plugins = self._plugins_resources(self)
 
-    def set_timeout(self, timeout):
+    def get_timeout(self, method=None):
+        """
+        Get timeout for given request method
+
+        :param method: Request method (eg. GET, POST, ..). If None, return
+            default timeout.
+        """
+        try:
+            timeout = self.timeout[method]
+        except KeyError:
+            logging.debug("Timeout for method %s not defined, using default" % method)
+            timeout = self.timeout[None]
+
+        if timeout == 0:
+            # Return None instead of 0 as it's not accepted by requests as
+            # no timeout
+            timeout = None
+        return timeout
+
+    def set_timeout(self, timeout, method=None):
         """
         Set the timeout for any connection, the timeout is the requests module
         timeout (for conneciton inactivity rather than request total time)
 
         :param timeout: Timeout in seconds for the connection inactivity
+        :param method: Request method (eg. GET, POST, ..). If None, return
+            default timeout.
         """
-        self._req_params['timeout'] = timeout
+        if method and timeout is None:
+            # timeout is None, so we want to use default timeout for this
+            # method
+            try:
+                self.timeout.pop(method)
+            except KeyError:
+                pass
+        else:
+            self.timeout[method] = timeout
```

Comment:
what if you want timeout by default except POST?

I think you should always set it like this:
    self.timeout[method] = timeout or None

have a default:
    self.timeout['DEFAULT'] = None

and when getting it just:
    self.timeout.get(method, self.timeout['DEFAULT'])

that way you can disable the timeout for a certain method while having it enabled by default. The drawback is that it requires another function to unset a method timeout:
    def unset_timeout(self, method):
        try:
            self.timeout.pop(method)
        except KeyError:
            pass